File: 프로그래머스/4/43236. 징검다리/징검다리.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int solution(int distance, vector<int> rocks, int n) {
    int answer = 0;
    sort(rocks.begin(), rocks.end());
    
    int left = 0;
    int right = distance;
    while (left <= right) {
        int mid = (left + right) / 2;
        
        int prev = 0;
        int removed = 0;
        for(auto r: rocks) {
            if (r - prev < mid) {
                removed++; 
            } else {
                prev = r;
            }
        }
        
        if (distance - prev < mid) {
            removed++;
        }
        
        if (removed > n) {
            right = mid - 1;
        } else {
            left = mid + 1;
            answer = mid;
        }
    }
    return answer;
}

File: 프로그래머스/4/43236. 징검다리/징검다리.java
import java.util.*;

class Solution {
    public int solution(int distance, int[] rocks, int n) {
        int answer = 0;
        int left = 0, right = distance;
        Arrays.sort(rocks);
        while (left <= right) {
            int mid = (left+right) / 2;
            
            int remove = 0;
            int prev = 0;
            for(int rock: rocks) {
                int dist = rock - prev;
                if (dist < mid) {
                    remove++;
                } else {
                    prev = rock;
                }
            }
            if (distance - prev < mid) remove++;
            if(remove > n) {
                right = mid-1;
            } else {
                answer = mid;
                left = mid +1;
            }
        }
        return answer;
    }
}